Loading wheel when storage is being accessed

andrayantelo opened the pull request:
Added a customEvent object on CheckIt. Pass it to
the storage object. Added startWork and endWork methods
in storage object.
Resolves issue #94

andrayantelo committed:
[WIP] Loading wheel when storage is being accessed
Added a customEvent object on CheckIt. Pass it to
the storage object. Added startWork and endWork methods
in storage object.

andrayantelo committed:
fixed two typos

andrayantelo committed:
Added startWork and endWork to one storage method
Added startWork and endWork to the method getAllCalendarIds
of the storage object in the way I think it is supposed to be used.
Added print statements so that I know what is actually getting run
and what isn't. Added dispatchEvent to startWork method but I don't know
where I'm supposed to be actually listening for events.
The activeCall counter in the storage object appears to be incrementing
and decrementing appropriately for the calls to storage that occur
when getting allCalendarIds.

andrayantelo committed:
Added a div for the loading wheel
It is the first child of the contentWrapper div

andrayantelo committed:
Added displayLoadingWheel and hideLoadingWheel
They are methods on the checkit object.

andrayantelo committed:
Added CSS for loadingWheel
Made it so that it appears in the middle of the webpage
and also set it's position as fixed.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -69,6 +72,19 @@ function CheckIt() {
+};
+
+CheckIt.prototype.hideLoadingWheel = function() {
+    spinner.stop();
```
This is an error.

`spinner` from `displayLoadingWheel` is not in scope here.

andrayantelo replied:
Do I make spinner a global variable? Or maybe a checkit variable?

andrayantelo replied:
Made it a checkit variable.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -808,32 +826,63 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
+    self.activeCalls = 0;
     
-    var toKey = function(id) {
+     var toKey = function(id) {
```
ws

andrayantelo replied:
Fixed.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -808,32 +826,63 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
+        self.activeCalls += 1;
+        
+        // Will dispatch the event backgroundActivityChange 
+        console.log("dipatching backgroundActivityChange");
```
I think you only want to do this when `activeCalls` is nonzero. The event handler will have no way of knowing this the way it is currently written.

You *could* include the current value of `activeCalls` as data on the event object, but I don't think that's necessary for what you're trying to do.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -808,32 +826,63 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
+        .then( function() {
+            
+            self.endWork();
+        }.bind(this))
```
Binding to `this` won't break anything here, but realize that you're mixing up two different strategies for dealing with the problem. On the one hand, you bound `self` to the store object and are using it in all the methods by closing over it. On the other hand, you're binding `this` to (what will probably be) that same object. You can certainly do both, but it's a little confusing, as it implies to the reader (that's you as soon as you forget all these details) that `this` and `self` are probably different. Otherwise why go through the trouble of having both?

johntyree replied:
You have to manually pass along the arguments if you want them to appear in future calls to `.then()`.

    .then(function(ids) {
        self.endWork();
        return ids;
    })

johntyree replied:
I suggest also passing a string to `startWork` and `endWork` so that you can log where they were called from, otherwise it's difficult to track why things are happening the way are.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -808,32 +826,63 @@ Calendar.prototype.fillCalendar = function(monthObjectsArray) {
+        }.bind(this))
+        .catch( function() {
+            
+            self.endWork();
```
The fact that you had to remember to duplicate this for both cases is precisely why Python has the `try ... except ... finally` syntax. https://docs.python.org/2.7/tutorial/errors.html#defining-clean-up-actions

johntyree replied:
Don't forget to pass along the error message here as well.

andrayantelo replied:
What am I duplicating? And also what error message?

johntyree replied:
You had to duplicate the call to `endWork()` in two places (in `.then()` and in `.catch()`) because you don't know which of the two will happen.

The error message is the one that is passed to this function as an argument. Whatever you called `reject` with.

johntyree commented on public/index.html:
```diff
@@ -144,6 +144,13 @@
 <div id="contentWrapper" class="container-fluid">
+
+
+    <div id="loadingWheel">
```
Why not attach the event to this element instead of `document` or `window` ?  Not sure how that affects dispatching, it might not be worth it.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -35,9 +35,12 @@ function CheckIt() {
     
     var monthObjects;
     
+    this.backgroundActivityChange = new CustomEvent('backgroundActivityChange');
```
I'm not sure what happens if you reuse a custom event, but I believe the typical way to do this is have the thing that will dispatch the event create a new one every time. This lets the dispatcher attach information to the event if necessary, for example, the number of active calls.

What you need to remember is the *type* of the event, which is just the string. That's all the listener needs to know in order to catch the event.

johntyree replied:
And actually... after playing around with this for awhile and reading the docs, it occurs to me that no one else does it this way. Usually the dispatcher just provides a function can call if you want to hook a listener up. We already see this with `firebase.auth.onAuthStateChanged` for example. I'd do it that way.

Remember that you'll probably end up with more than one thing that needs to react to busy-state changes, so it's not enough to just remember a single callback. There will be many of them and you'll need to invoke them all.

I have some example code that does this, but you should give it a try first.

andrayantelo replied:
What did you google for when you looked around to see what other people do?

andrayantelo replied:
Hmm.. can you be more specific with "Usually the dispatcher just provides a function can call if you want to hook a listener up." How does it work? Or point me in the direction where I can read about it.

johntyree replied:
The most recent and obvious example is `firebase.auth.onAuthStateChanged()`. You want to basically the same thing here. You have an object (`auth` or the `store`) that has some kind of hidden internal state that you don't really care about but you want to know when something important happens, such as logging in or out for `auth` or starting and finishing work for the `store`.

johntyree reviewed (changes requested):
I suggest merging your jitter branch into master and then merging master into this branch before continuing. The fact that everything is happening instantaneously is making it impossible to actually see the spin wheel because there's never any background activity.

You can always set the jitter to 0 if you want it to be fast again.

andrayantelo committed:
Merge branch 'master' into loadingWheel

andrayantelo committed:
Added string argument to startWork and endWork
Got rid of white space, passed ids variable to then
function

andrayantelo committed:
Got rid of custom event backgroundActivityChange
No longer pass it to storage object, does not get
dispatched inside of storage method startWork

andrayantelo committed:
Made spinner a variable on checkit object

andrayantelo committed:
Included setUpLoadingWheel methond on Checkit
setUpLoadingWheel creates the Spinner object.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -820,6 +872,27 @@ var LocalCalendarStorage = function(params) {
+        }
+    };
+    
+    self.endWork = function(accesor) {
```
`s/accessor/processName`

johntyree commented on public/calendar_reboot.js:
```diff
@@ -820,6 +872,27 @@ var LocalCalendarStorage = function(params) {
+        self.activeCalls += 1;
+        
+        // Will dispatch the event backgroundActivityChange 
+        if (self.activeCalls > 0) {
```
Think about how the interaction between the store and the listeners will work. Is the store going to invoke the callbacks every time an operation starts? If so, you'll need to include the value of `activeCalls` somehow so that the listener can decide whether you're just starting to work or are doing more than one thing at a time.

What I mean is that the transition from 0 to 1 is semantically more important than say, going from 5 to 6, right? How will listeners know?

andrayantelo replied:
Yeah, I understand. You mean if I had a customEvent for example I'd add activeCalls to it's details right? So, that you can access it. Except I'm not using customEvents anymore. 

So, it would be ... in the function call... that you are listening to?

johntyree commented on public/calendar_reboot.js:
```diff
@@ -69,6 +71,52 @@ function CheckIt() {
 
 };
 
+CheckIt.prototype.setUpLoadingWheel = function() {
```
Do you call this more than once? Why not just make it part of the constructor?

```javascript
this.spinner = new Spinner({
    lines: 13
  , length: 28
  ...
});
```

johntyree commented on public/index.html:
```diff
@@ -144,6 +144,13 @@
 <div id="contentWrapper" class="container-fluid">
+
+
+    <div id="loadingWheel">
```
I think you're going to discover that planting a giant spinner wheel in the middle of the UI everytime you touch storage is a bit too invasive.

johntyree replied:
You have two divs with id `loadingWheel` ?

johntyree commented on public/calendar_reboot.js:
```diff
@@ -69,6 +71,52 @@ function CheckIt() {
+    var target = document.getElementById("loadingWheel");
+    // Hardcoded for now.
+    this.$buildFormAccordion.attr('hidden', true);
+    this.clearPage();
```
So... every time I click a checkmark you're going to clear the page?

andrayantelo replied:
Yeah, I guess not. lol
I was only considering the time when a calendar is being loaded.

andrayantelo replied:
I guess I won't need to be hiding the build form every time either.

andrayantelo committed:
setUpLoadingWheel's function part of constructor
kept opts variable separate for now.

andrayantelo committed:
Spinner needs plugin for jQuery
Using vanilla JavaScript instead.

andrayantelo committed:
Added argument to displayLoadingWheel
Argument for elementId so that you can specify where you want
to stick the loading Wheel.

andrayantelo committed:
unindented lines

andrayantelo committed:
checkit method that inserts a div where you need
with an id that you provide it.

andrayantelo committed:
Added onActivityChanged to checkit and store
Checkit's has a string explaining what it will do.
Store kind of does but not really  because I'm still
trying to understand how this is going to work.

andrayantelo committed:
Pass checkit's onActivityChange to store object

andrayantelo committed:
Pass checkit's onActivityChanged to store's
as an argument and store it as a property on store.
Run it in startWork and endWork methods which are currently
only being used in the allCalendarIds method of store.

andrayantelo committed:
Display and hide loading wheel in checkit method
the number of active calls is passed to checkit's onActivityChanged method. If the active calls is equal to zero the loadingWheel is hidden. If it is greater than zero (meaning there are one or more active calls occuring, the loading wheel should be shown. Right now the loading wheel is being shown in the middle of the page, but that will likely change to a smaller area until I decide whether or not I want the loading wheel's location to depend on what active call is currently running.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -207,6 +208,8 @@ CheckIt.prototype.onActivityChanged = function(storageObj) {
     // Passing storageObj as argument to have access to activity calls, which
```
Just pass the activity level directly if you want to do this. If you really want to access the entire store, you could have gotten it via `this.store` anyway.

johntyree reviewed (commented):
So far so good.

For bonus points, `store.onActivityChanged` should be callable multiple times to support having more than one callback function.

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -852,21 +971,28 @@ var LocalCalendarStorage = function(params) {
         //store the state in localStorage
 
         var stateP = new Promise(function(resolve, reject) {
+            //self.startWork();
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -852,21 +971,28 @@ var LocalCalendarStorage = function(params) {
             storeInLocalStorage(toKey(calendarObj.state.uniqueId), calendarObj.state);
             jitter(resolve);
         });
+        // When do I "endWork" here? in the stateP promise?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -852,21 +971,28 @@ var LocalCalendarStorage = function(params) {
+        
         var idsP = self.getAllCalendarIds()
             .then(function (allCalendarIds) {
+                // start work here?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -852,21 +971,28 @@ var LocalCalendarStorage = function(params) {
+                // start work here?
                 allCalendarIds[calendarObj.state.uniqueId] = calendarObj.state.title;
                 storeInLocalStorage(toKey(allCalendarIdsKey), allCalendarIds);
+                // end work here?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -852,21 +971,28 @@ var LocalCalendarStorage = function(params) {
             })
             .catch(function () {
                 console.log("No previous calendars in storage");
+                // start work?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -852,21 +971,28 @@ var LocalCalendarStorage = function(params) {
                 var allCalendarIds = {};
                 allCalendarIds[calendarObj.state.uniqueId] = calendarObj.state.title;
                 storeInLocalStorage(toKey(allCalendarIdsKey), allCalendarIds);
+                // end work?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -886,6 +1012,8 @@ var LocalCalendarStorage = function(params) {
 
         return self.getAllCalendarIds()
             .then(function(allCalendarIds) {
+                // startwork?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -894,22 +1022,28 @@ var LocalCalendarStorage = function(params) {
                 // Remove active status from the calendar
                 removeFromLocalStorage(toKey(current_active_calendar));
+                
+                // end work?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -894,22 +1022,28 @@ var LocalCalendarStorage = function(params) {
             }.bind(this))
             .catch(function () {
                 console.log("Unable to remove calendar");
+                // end work?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -894,22 +1022,28 @@ var LocalCalendarStorage = function(params) {
     self.loadById = function(calendarObjId) {
         // Return a promise to a calendar object using its Id
         return new Promise( function(resolve, reject) {
+            //self.startWork(); // ?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -894,22 +1022,28 @@ var LocalCalendarStorage = function(params) {
             
             if (calendar !== null) {
                 jitter(resolve, calendar);
+                //self.endWork(); // ?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -894,22 +1022,28 @@ var LocalCalendarStorage = function(params) {
+                //self.endWork(); // ?
             }
             else jitter(reject, "Calendar not found");
+            //self.endWork(); //?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -918,15 +1052,18 @@ var LocalCalendarStorage = function(params) {
         // Return a promise to the active_calendar id from storage
         
         return new Promise( function(resolve, reject) {
+            //self.startWork() //?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -918,15 +1052,18 @@ var LocalCalendarStorage = function(params) {
                 
                 // Signal that the promise succeeded and make the value ready to go. 
                 jitter(resolve, activeCalendarId);
+                //self.endWork() //?
```
?

andrayantelo commented on public/calendar_reboot.js:
```diff
@@ -918,15 +1052,18 @@ var LocalCalendarStorage = function(params) {
             }
             else {
                 jitter(reject, "Not found");
+                //self.endwork() //?
```
?

johntyree replied:
I explained the rational behind when you should or should not call `endWork` below.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -828,21 +933,35 @@ var LocalCalendarStorage = function(params) {
         })
+        .then( function(ids) {
+            console.log("getAllCalendarIds then function running");
+            self.endWork(); // should I actually be ending work here?
```
Yes. The purpose of these is to indicate whether or not work is *still happening* and will finish sometime in the future. If you call `startWork` without calling `endWork` you will appear to be stuck in the "activity" state forever. You need to make sure that no matter what happens, `startWork` and `endWork` get called exactly the same number of times.

Also, don't forget to remove this comment.

andrayantelo committed:
Added startWork and endWork to jitter function

andrayantelo committed:
changed CSS and positioning for loadingWheel div

johntyree commented on public/calendar_reboot.js:
```diff
@@ -69,6 +97,24 @@ function CheckIt() {
+           // where you want to place the loading wheel.
+    
+    // Needs a plugin to use jQuery.
+    var target = document.getElementById(elementId);
```
You can still user jQuery here and just grab the element out of it before you pass it to the spinner.
I think it's `this.spinner.spin($(selector).get());`.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -820,12 +887,48 @@ var LocalCalendarStorage = function(params) {
+    self.endWork = function() {
+        // Will decrement the counter and maybe fire an event.
+        
+        self.activeCalls -= 1;
```
Can you add an assertion that `activeCalls >= 0` here? I have a feeling it will be tricky to make sure there is exactly one invocation of `endWork` for every `startWork`.

johntyree replied:
Oh I see the log message now. I'd argue it should be `console.error` if `activeCalls < 0`.

andrayantelo committed:
got rid of extra loadingWheel div, select loadingWheel with jQuery for spinner

andrayantelo committed:
Got rid of some print statements.

johntyree commented on public/calendar_reboot.js:
```diff
@@ -104,9 +104,8 @@ CheckIt.prototype.displayLoadingWheel = function(elementId) {
     // Parameters: elementId 
            // A string pertaining to the id of the element
            // where you want to place the loading wheel.
+    var target = $('#loadingWheel').get()[0];
```
`elementId` not `#loadingWheel`

johntyree commented on public/calendar_reboot.js:
```diff
@@ -899,7 +897,7 @@ var LocalCalendarStorage = function(params) {
         
         self.activeCalls += 1;
-        console.log("store's startWork is running, activeCalls = " + self.activeCalls);
+        self.activeCalls += 1;
```
What's going on here? You increment twice?

andrayantelo committed:
got rid of typo and some print statements

johntyree commented on public/calendar_reboot.js:
```diff
@@ -911,8 +909,9 @@ var LocalCalendarStorage = function(params) {
+        
         self.activeCalls -= 1;
-        console.log("store's endWork is running, activeCalls = " + self.activeCalls);
+        if (self.activeCalls < 0) console.error("No work has been started");
```
It is best not to get into the habit of leaving out the curly braces. The benefit is very small and the likelihood of introducing bugs accidentally is very high. https://www.reddit.com/r/programming/comments/22ghj1/the_heartbleed_bug/

andrayantelo committed:
Added curly braces
